### PR TITLE
Dockerification

### PR DESCRIPTION
**Pull request by acheronw**
Manager webapp with autorunner function and a single does-it-all script!

### CODE REVIEW

File: app/seed_db.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
```
Comment (DavidNemeskey) [inline app/seed_db.py:1]: Why is this file needed?

File: app/main.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+from sqlalchemy.orm import Session
+
+
+from . import crud, models, seed_db, schemas
+from .config import config
+from .database import SessionLocal, engine
+
+
+# TODO this should be replaced by proper migrations using the Alembic library
+models.Base.metadata.create_all(bind=engine)
```
Comment (DavidNemeskey) [inline app/main.py:25]: What does this line do? Please document it.

File: app/main.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+    db = SessionLocal()
+    try:
+        yield db
+    finally:
+        db.close()
+
+
+# TODO this is for development purposes. Will we need seeding for production?
+@app.on_event("startup")
+def dev_seed():
```
Comment (DavidNemeskey) [inline app/main.py:51]: Isn't this duplicated  by `seed_db.py`?

File: app/main.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    context = {"request": request, "pipeline": db_pipeline}
+    return templates.TemplateResponse("view_pipeline.html", context)
+
+
+@app.get("/edit_pipeline/{pipeline_id}", response_class=HTMLResponse)
+async def edit_pipeline(pipeline_id: int,
+                        request: Request,
+                        db: Session = Depends(get_db)
+                        ):
+    """Controller to render the edit pipeline form."""
```
Comment (DavidNemeskey) [inline app/main.py:318]: "Controller" as in MVC? Wouldn't that make this a "view"?

File: app/main.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+    # Convert nested Dicts and replace empty strings with Nones:
+    if form_data["params"]:
+        form_data["params"] = loads(form_data["params"])
+    else:
+        form_data["params"] = None
+    if form_data["steps"]:
+        form_data["steps"] = loads(form_data["steps"])
+    else:
+        form_data["steps"] = None
+    if not form_data["prereq_pipe"]:
```
Comment (DavidNemeskey) [inline app/main.py:350]: `setdefault`

File: cc_corpus/download.py [hunk 6]
```diff
@@ -67,34 +70,40 @@ def download_ranges(url: str,
...
             continue
     else:
         raise DownloadError(f'Could not download ranges from URL {url}.')
 
 
 def download_warc_ranges(
-    warc_file_name: str,
-    offsets_and_lengths: list[tuple[int, int]],
-    retry_left: int
+        warc_file_name: str,
```
Comment (DavidNemeskey) [inline cc_corpus/download.py:80]: The default indent in Python is 4 spaces, not 8.

File: cc_corpus/download.py [hunk 7]
```diff
@@ -67,34 +70,40 @@ def download_ranges(url: str,
...
+        delay_period=delay
     )
 
 
 def download_index_range(
-    index_file_url: str,
-    offset: int,
-    length: int,
-    retry_left: int
+        index_file_url: str,
```
Comment (DavidNemeskey) [inline cc_corpus/download.py:97]: Again, 4 spaces.

File: scripts/api_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        format='%(asctime)s - %(threadName)-10s)- %(levelname)s - %(message)s',
+        level=logging.INFO
+    )
+
+    logging.info(f"Script {sys.argv[3]} (id: {step_id}) - Starting with: {sys.argv[2:]}")
+    results = subprocess.run(sys.argv[3:])
+    if results.returncode == 0:
+        logging.info(f"Script {sys.argv[3]} (id: {step_id}) - Successfully executed")
+        # TODO this url shoud not be hardwired:
+        success_url = f"http://127.0.0.1:8000/completed/{step_id}"
```
Comment (DavidNemeskey) [inline scripts/api_wrapper.py:32]: Please make this URL not hardwired :smile:

File: scripts/api_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+"""
+
+import logging
+import sys
+import requests
+import subprocess
+
+
+def main():
+    step_id = sys.argv[2]
```
Comment (DavidNemeskey) [inline scripts/api_wrapper.py:18]: `argparse`

File: scripts/download_pages.py [hunk 10]
```diff
@@ -307,20 +307,22 @@ def main():
...
     else:
         print('Sorting index...')
-        step1(args.input_pattern, ranges_dir)
+        create_sorted_index(args.input_dir, ranges_dir)
 
     print('Downloading pages...')
-    step2(ranges_dir, args.processes, args.index_output_dir,
-          args.data_output_dir, args.error_file, args.retry, args.chunksize,
-          args.out_filename, args.padding, args.ext)
+    download_collected_ranges(ranges_dir, args.processes, args.index_output_dir,
```
Comment (DavidNemeskey) [inline scripts/download_pages.py:323]: Aren't these lines too long?

File: scripts/pipe_that_line.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    ppp = set()
+    pipes = config["pipelines"]
+    for pipe in pipes.values():
+        ppp.update(pipe["params"])
+    return ppp
+
+
+def parse_arguments():
+    parser = ArgumentParser(description=__doc__)
+    parser.add_argument('--pipe', '-p', required=True,
```
Comment (DavidNemeskey) [inline scripts/pipe_that_line.py:29]: Should be a positional argument.

File: scripts/filter_index.py [hunk 12]
```diff
@@ -42,8 +43,8 @@
 
 def parse_arguments():
     parser = ArgumentParser(description=__doc__)
-    parser.add_argument('input_dir', help='the input directory.')
-    parser.add_argument('output_dir', help='the output directory.')
+    parser.add_argument('--input_dir', '-i', help='the input directory.')
```
Comment (DavidNemeskey) [inline scripts/filter_index.py:46]: Aren't these required arguments? I would say they should stay positional then.

File: scripts/pipe_that_line.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+                      f"config files.")
+    params = {}
+    for param in pipe_params:
+        param_value = getattr(args, param, None)
+        if param_value:
+            params[param] = param_value
+        else:
+            logging.error(f"Pipeline type {args.pipe} requires parameter "
+                          f"{param} but it was not supplied.")
+            raise KeyError
```
Comment (DavidNemeskey) [inline scripts/pipe_that_line.py:67]: I am not sure about raising exceptions in `main`. `sys.exit`?

File: app/crud.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+                if value.get('no_batch_in_path'):
+                    further_params += " -" + key + " " + dir_head + value[key]
+                else:
+                    further_params += " -" + key \
+                                      + " " + dir_head + value[key] + dir_tail
+
+    db_step.further_params = further_params
+    db.add(db_step)
+    db.commit()
+    db.refresh(db_step)
```
[the remark this answers is not in the document]
Comment (acheronw) [reply app/crud.py:84]: I checked it and you are right, we don't need it here. 

Default behaviour of commit() is that it releases the object, so a new one will be fetched anyway, no need for refresh().

By the way: we need the reloaded object because it is returned by this method and rendered as a response.

File: app/crud.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+    return db_step
+
+
+def update_step(db: Session, step: schemas.StepUpdate):
+    """
+    Updates a Step in the DB
+    Fields that are missing from the update object are left unchanged.
+    """
+    # get existing data from the DB:
+    db_step = db.query(models.Step).\
```
[the remark this answers is not in the document]
Comment (acheronw) [reply app/crud.py:94]: It's ugly either way. 

You know what? They can dig that 80 character per line rule!

File: app/crud.py [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+    db.query(models.Step).filter(models.Step.id == step_id).delete()
+    db.commit()
+
+
+def get_pipelines(db: Session, skip: int = 0, limit: int = 100):
+    """Fetches Pipeline objects from the DB."""
+    return db.query(models.Pipeline).offset(skip).limit(limit).all()
+
+
+def get_pipelines_by_status(db: Session, status: str):
```
[the remark this answers is not in the document]
Comment (acheronw) [reply app/crud.py:117]: Yes, you are right, these should be merged.

BTW: the limit+offset is there to support pagination, but on the frontend pagination is not implemented yet.

With 3 optional paramteres I think the filtering dictionary would be an overkill. 
If, in the future, we decide to pimp up the frontend even more (which does not seem too likely) and add further filters there, like filtering by timestamps and such, then we could reconsider using a filtering dictionary.

File: app/crud.py [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+                elif step.status == "failed":
+                    # We cannot progress with this pipeline:
+                    print(f"--Step #{step.id} failed.")
+                    encountered_failure = True
+                    break
+                else:
+                    # We have to run this step:
+                    step.run_script()
+                    step.status = "running"
+                    db.commit()
```
[the remark this answers is not in the document]
Comment (acheronw) [reply app/crud.py:278]: Commit:
We have to commit that change in step status into the DB right there.

Because when the background execution of the script in run_script finishes, it will call back to the API and will set the step status from "running" to either "complete" or "failed".

Try-except-finally: I don't think so. Since this is a manager webapp, I think we are fine with crashing if there is a DB-related error. 
The worst thing that could happen (at least what came to my mind) is that a background task completes a step and there is no app waiting for its completed signal, and thus the fact of the completion is not noted in the DB, so we have to redo it. 
But if there is an error in the DB with updating records, then we could hardly avoid that problem with a try-except-finally.

File: app/main.py (hunk 3 above)
Comment (acheronw) [reply app/main.py:51]: No, we discussed it.

File: app/main.py (hunk 4 above)
Comment (acheronw) [reply app/main.py:318]: Nope, this is the "controller" in MVC, because it controls the flow of the app and of the data. The "view" part is the templates/edit_pipeline.html, while the "model" part is what happens in the crud.py, models.py and schemas.py.

File: app/main.py (hunk 5 above)
Comment (acheronw) [reply app/main.py:350]: I am not dealing with missing keys here. I am replacing existing, empty string values with None values, so those fields will be left unchanged when updating the record instead of being replaced with emtpy strings.

So I don't think setdefault would help me with that, would it?

File: cc_corpus/download.py (hunk 6 above)
Comment (acheronw) [reply cc_corpus/download.py:80]: PEP 8 style guide:

# Correct:
# Add 4 spaces (an extra level of indentation) to distinguish arguments from the rest.
def long_function_name(
        var_one, var_two, var_three,
        var_four):
    print(var_one)

# Wrong:
def long_function_name(
    var_one, var_two, var_three,
    var_four):
    print(var_one)

So I kept the 8 spaces of indentation, but moved the closing bracket to the end of the last parameter line, instead of starting a line of its own.

File: cc_corpus/download.py (hunk 7 above)
Comment (acheronw) [reply cc_corpus/download.py:97]: Same thing, kept the 8 space indent, moved closing parenthes up one line.

File: scripts/api_wrapper.py (hunk 8 above)
Comment (acheronw) [reply scripts/api_wrapper.py:32]: I spent hours trying to fix this.

The problem is that to dynamically get the current IP address of the webapp itself you have to query that from the fastapi object. 
That is currently created in our main.py:  
`app = FastAPI(title="CC Corpus manager")`

Trying to give access to that object or any variable created in the main.py leads to circular imports error. I tried to work around that by moving it to a third python file, but I still could not avoid the circular imports error.

Just now a new idea came to my mind: maybe it should write its own ip address into the DB on startup, and I could fetch the value from there, and thus avoid the circular imports error...

File: scripts/api_wrapper.py (hunk 9 above)
Comment (acheronw) [reply scripts/api_wrapper.py:18]: When I first wrote this code it seemed easier to just use sys.argv, since most of the args will be passed down to the next script and there is an unknown number of them.

Then things got a bit more complicated. I'm still not sure is it worth using argparse...

File: scripts/filter_index.py (hunk 12 above)
Comment (acheronw) [reply scripts/filter_index.py:46]: You are right, they are required.

But I got rid off positional arguments and instead adopted unified shortcodes (always -i and -o, and always -P for process if that is applicable) for these arguments, so the manager app can put together the proper arguments for each script easily.

But I will add the required=True flag, that's a good idea.

File: scripts/pipe_that_line.py (hunk 13 above)
Comment (acheronw) [reply scripts/pipe_that_line.py:67]: Well, raising a KeyError seemed more informative than simply exiting the code...

File: app/main.py [hunk 18]
```diff
@@ -0,0 +1,380 @@
...
+
+@app.post("/update_pipeline/", response_class=HTMLResponse)
+async def update_pipeline_from_form(request: Request,
+                                    db: Session = Depends(get_db),
+                                    ):
+    """Controller to process the response from the edit pipeline form."""
+    form_data = await request.form()
+    form_data = jsonable_encoder(form_data)
+
+    # Convert nested Dicts and replace empty strings with Nones:
```
Comment (DavidNemeskey) [inline app/main.py:345]: Explain why (FastAPI)
